### ISSUE TITLE
certbot 1.10.1 fails to revoke ecdsa certificates using --cert-key

### ISSUE THREAD

**Issue by commonism** [certbot/certbot#8569]
Hi,

I'm with python3.8, certbot via pip, localhost ca.

certbot wants to sign the revocation with the certificate key.
Current josepy lacks support for ecdsa, resulting in:
```
/tmp/cb/bin/certbot --non-interactive -c tmp/CertBotRSA2048EC256CA/certbot.ini --server http://localhost:8123/directory revoke --cert-path tmp/CertBotRSA2048EC256CA/etc/letsencrypt/live/f.localca2.test.de/cert.pem --key-path tmp/CertBotRSA2048EC256CA/etc/letsencrypt/live/f.localca2.test.de/privkey.pem
Saving debug log to /tmp/certbot/logs/letsencrypt.log
An unexpected error occurred:
TypeError: __init__() takes exactly the following arguments:  (key given)
Please see the logfiles in /tmp/certbot/logs for more details.
```

Logfile content:
```
2020-12-31 11:30:09,465:DEBUG:certbot._internal.main:certbot version: 1.10.1
2020-12-31 11:30:09,466:DEBUG:certbot._internal.main:Arguments: ['--non-interactive', '-c', 'tmp/CertBotRSA2048EC256CA/certbot.ini', '--server', 'http://localhost:8123/directory', '--cert-path', 'tmp/CertBotRSA2048EC256CA/etc/letsencrypt/live/f.localca2.test.de/cert.pem', '--key-path', 'tmp/CertBotRSA2048EC256CA/etc/letsencrypt/live/f.localca2.test.de/privkey.pem']
2020-12-31 11:30:09,466:DEBUG:certbot._internal.main:Discovered plugins: PluginsRegistry(PluginEntryPoint#manual,PluginEntryPoint#null,PluginEntryPoint#standalone,PluginEntryPoint#webroot)
2020-12-31 11:30:09,478:DEBUG:certbot._internal.log:Root logging level set at 20
2020-12-31 11:30:09,478:INFO:certbot._internal.log:Saving debug log to /tmp/certbot/logs/letsencrypt.log
2020-12-31 11:30:09,479:DEBUG:certbot._internal.main:Revoking /srv/home/common/workspace/acme-broker/tests/tmp/CertBotRSA2048EC256CA/etc/letsencrypt/live/f.localca2.test.de/cert.pem using cert key /srv/home/common/workspace/acme-broker/tests/tmp/CertBotRSA2048EC256CA/etc/letsencrypt/live/f.localca2.test.de/privkey.pem
2020-12-31 11:30:09,479:DEBUG:certbot._internal.log:Exiting abnormally:
Traceback (most recent call last):
  File "/tmp/cb/bin/certbot", line 8, in <module>
    sys.exit(main())
  File "/tmp/cb/lib/python3.8/site-packages/certbot/main.py", line 15, in main
    return internal_main.main(cli_args)
  File "/tmp/cb/lib/python3.8/site-packages/certbot/_internal/main.py", line 1412, in main
    return config.func(config, plugins)
  File "/tmp/cb/lib/python3.8/site-packages/certbot/_internal/main.py", line 1103, in revoke
    key = jose.JWK.load(config.key_path[1])
  File "/tmp/cb/lib/python3.8/site-packages/josepy/jwk.py", line 119, in load
    return jwk_cls(key=key)
  File "/tmp/cb/lib/python3.8/site-packages/josepy/json_util.py", line 217, in __init__
    super(JSONObjectWithFields, self).__init__(
  File "/tmp/cb/lib/python3.8/site-packages/josepy/util.py", line 150, in __init__
    raise TypeError(
TypeError: __init__() takes exactly the following arguments:  (key given)
2020-12-31 11:30:09,480:ERROR:certbot._internal.log:An unexpected error occurred:
2020-12-31 11:30:09,480:ERROR:certbot._internal.log:TypeError: __init__() takes exactly the following arguments:  (key given)
```

Using the josepy patches https://github.com/certbot/josepy/pull/51 to provide ECDSA results in …

```
…
Saving debug log to /tmp/certbot/logs/letsencrypt.log
An unexpected error occurred:
AssertionError
Please see the logfiles in /tmp/certbot/logs for more details.
```

```
2020-12-31 11:41:46,191:DEBUG:certbot._internal.main:certbot version: 1.10.1
2020-12-31 11:41:46,191:DEBUG:certbot._internal.main:Arguments: ['--non-interactive', '-c', 'tmp/CertBotRSA2048EC256CA/certbot.ini', '--server', 'http://localhost:8123/directory', '--cert-path', 'tmp/CertBotRSA2048EC256CA/etc/letsencrypt/live/f.localca2.test.de/cert.pem', '--key-path', 'tmp/CertBotRSA2048EC256CA/etc/letsencrypt/live/f.localca2.test.de/privkey.pem']
2020-12-31 11:41:46,191:DEBUG:certbot._internal.main:Discovered plugins: PluginsRegistry(PluginEntryPoint#manual,PluginEntryPoint#null,PluginEntryPoint#standalone,PluginEntryPoint#webroot)
2020-12-31 11:41:46,202:DEBUG:certbot._internal.log:Root logging level set at 20
2020-12-31 11:41:46,202:INFO:certbot._internal.log:Saving debug log to /tmp/certbot/logs/letsencrypt.log
2020-12-31 11:41:46,203:DEBUG:certbot._internal.main:Revoking /srv/home/common/workspace/acme-broker/tests/tmp/CertBotRSA2048EC256CA/etc/letsencrypt/live/f.localca2.test.de/cert.pem using cert key /srv/home/common/workspace/acme-broker/tests/tmp/CertBotRSA2048EC256CA/etc/letsencrypt/live/f.localca2.test.de/privkey.pem
2020-12-31 11:41:46,204:DEBUG:acme.client:Sending GET request to http://localhost:8123/directory.
2020-12-31 11:41:46,205:DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): localhost:8123
2020-12-31 11:41:46,208:DEBUG:urllib3.connectionpool:http://localhost:8123 "GET /directory HTTP/1.1" 200 303
2020-12-31 11:41:46,208:DEBUG:acme.client:Received response:
HTTP 200
Content-Type: application/json; charset=utf-8
Link: <http://localhost:8123/directory>; rel="index"
Replay-Nonce: 45dc74fdb9bb426986fbfd9eecfdb310
Cache-Control: no-store
Content-Length: 303
Date: Thu, 31 Dec 2020 10:41:46 GMT
Server: Python/3.8 aiohttp/3.6.3

{"newAccount": "http://localhost:8123/new-account", "newNonce": "http://localhost:8123/new-nonce", "newOrder": "http://localhost:8123/new-order", "revokeCert": "http://localhost:8123/revoke-cert", "meta": {"termsOfService": "https://tos.acmenoah.luis.uni-hannover.de", "externalAccountRequired": false}}
2020-12-31 11:41:46,209:DEBUG:certbot._internal.main:Reason code for revocation: 0
2020-12-31 11:41:46,209:DEBUG:acme.client:Requesting fresh nonce
2020-12-31 11:41:46,209:DEBUG:acme.client:Sending HEAD request to http://localhost:8123/new-nonce.
2020-12-31 11:41:46,211:DEBUG:urllib3.connectionpool:http://localhost:8123 "HEAD /new-nonce HTTP/1.1" 204 0
2020-12-31 11:41:46,211:DEBUG:acme.client:Received response:
HTTP 204
Link: <http://localhost:8123/directory>; rel="index"
Replay-Nonce: 552c6b3db96d4053916b1140eb5bd26c
Cache-Control: no-store
Content-Length: 0
Content-Type: application/octet-stream
Date: Thu, 31 Dec 2020 10:41:46 GMT
Server: Python/3.8 aiohttp/3.6.3


2020-12-31 11:41:46,211:DEBUG:acme.client:Storing nonce: 552c6b3db96d4053916b1140eb5bd26c
2020-12-31 11:41:46,212:DEBUG:acme.client:JWS payload:
b'{\n  "certificate": "MIICcDCCAVigAwIBAgIUd_RypQd7JDUd2Zk0v0IGagz3Sf0wDQYJKoZIhvcNAQELBQAwXDELMAkGA1UEBhMCREUxFTATBgNVBAgMDExvd2VyIFNheG9ueTEQMA4GA1UEBwwHSGFub3ZlcjEUMBIGA1UECgwLQWNtZSBCcm9rZXIxDjAMBgNVBAcMBUFCIENBMB4XDTIwMTIzMDEwMzAwMFoXDTIxMDEyOTEwMzAwMFowHTEbMBkGA1UEAwwSZi5sb2NhbGNhMi50ZXN0LmRlMFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEmTSaWScMRhmjhbg9oPpCwiO5WGJ6facr21q-MCRwQ7dcJeIC1adiEfbsWqFGB2xFxy3cc5ZepBNNPKxt0BjokqM0MDIwMAYDVR0RBCkwJ4ISZi5sb2NhbGNhMi50ZXN0LmRlghFmLmxvY2FsY2EudGVzdC5kZTANBgkqhkiG9w0BAQsFAAOCAQEARAhDyoTnQquUxBPw5i7PGuAQha1PZVkuEXrXZEw8ddfr6G8hoGzShGAI9nmffKj9nxsZleF_XcfEBrkgB0dvYa4t3NBnwb-RAjYHA5W49fdXyfiAJ1leldhM-zl_VKMZ3loDrb4rJwqjVZzVvLf7_85FK3_acDJlwT4zENWeZmIiY_Fw4SglW3B1FAbrmYsdNt--SBMcuDE0XgisgF3KzvGTOp77avwvWlw_u4yZXi9-xyjvfi6GMR1uNaKxYJ3GNftiimQH_S_f9MpM7-u0iXeR0PWbYlxL4KfmD0KlZ6PZN-ZVjsMpJHYBvzvfFGjr2MVQHnvjAv5SydeB6egcJA",\n  "reason": 0\n}'
2020-12-31 11:41:46,213:DEBUG:certbot._internal.log:Exiting abnormally:
Traceback (most recent call last):
  File "/tmp/cb/bin/certbot", line 8, in <module>
    sys.exit(main())
  File "/tmp/cb/lib/python3.8/site-packages/certbot/main.py", line 15, in main
    return internal_main.main(cli_args)
  File "/tmp/cb/lib/python3.8/site-packages/certbot/_internal/main.py", line 1412, in main
    return config.func(config, plugins)
  File "/tmp/cb/lib/python3.8/site-packages/certbot/_internal/main.py", line 1112, in revoke
    acme.revoke(jose.ComparableX509(cert), config.reason)
  File "/tmp/cb/lib/python3.8/site-packages/acme/client.py", line 938, in revoke
    return self.client.revoke(cert, rsn)
  File "/tmp/cb/lib/python3.8/site-packages/acme/client.py", line 780, in revoke
    return self._revoke(cert, rsn, self.directory['revokeCert'])
  File "/tmp/cb/lib/python3.8/site-packages/acme/client.py", line 225, in _revoke
    response = self._post(url,
  File "/tmp/cb/lib/python3.8/site-packages/acme/client.py", line 97, in _post
    return self.net.post(*args, **kwargs)
  File "/tmp/cb/lib/python3.8/site-packages/acme/client.py", line 1201, in post
    return self._post_once(*args, **kwargs)
  File "/tmp/cb/lib/python3.8/site-packages/acme/client.py", line 1211, in _post_once
    data = self._wrap_in_jws(obj, self._get_nonce(url, new_nonce_url), url, acme_version)
  File "/tmp/cb/lib/python3.8/site-packages/acme/client.py", line 1030, in _wrap_in_jws
    return jws.JWS.sign(jobj, **kwargs).json_dumps(indent=2)
  File "/tmp/cb/lib/python3.8/site-packages/acme/jws.py", line 51, in sign
    return super(JWS, cls).sign(payload, key=key, alg=alg,
  File "/tmp/cb/lib/python3.8/site-packages/josepy-1.3.0.dev0-py3.8.egg/josepy/jws.py", line 267, in sign
    cls.signature_cls.sign(payload=payload, **kwargs),))
  File "/tmp/cb/lib/python3.8/site-packages/josepy-1.3.0.dev0-py3.8.egg/josepy/jws.py", line 208, in sign
    assert isinstance(key, alg.kty)
AssertionError
2020-12-31 11:41:46,214:ERROR:certbot._internal.log:An unexpected error occurred:
2020-12-31 11:41:46,214:ERROR:certbot._internal.log:AssertionError
```

removing the assert does not improve things …

```
2020-12-31 11:52:55,568:DEBUG:acme.client:JWS payload:
2020-12-31 11:52:55,569:DEBUG:certbot._internal.log:Exiting abnormally:
Traceback (most recent call last):
  File "/tmp/cb/bin/certbot", line 8, in <module>
    sys.exit(main())
  File "/tmp/cb/lib/python3.8/site-packages/certbot/main.py", line 15, in main
    return internal_main.main(cli_args)
  File "/tmp/cb/lib/python3.8/site-packages/certbot/_internal/main.py", line 1412, in main
    return config.func(config, plugins)
  File "/tmp/cb/lib/python3.8/site-packages/certbot/_internal/main.py", line 1112, in revoke
    acme.revoke(jose.ComparableX509(cert), config.reason)
  File "/tmp/cb/lib/python3.8/site-packages/acme/client.py", line 938, in revoke
    return self.client.revoke(cert, rsn)
  File "/tmp/cb/lib/python3.8/site-packages/acme/client.py", line 780, in revoke
    return self._revoke(cert, rsn, self.directory['revokeCert'])
  File "/tmp/cb/lib/python3.8/site-packages/acme/client.py", line 225, in _revoke
    response = self._post(url,
  File "/tmp/cb/lib/python3.8/site-packages/acme/client.py", line 97, in _post
    return self.net.post(*args, **kwargs)
  File "/tmp/cb/lib/python3.8/site-packages/acme/client.py", line 1201, in post
    return self._post_once(*args, **kwargs)
  File "/tmp/cb/lib/python3.8/site-packages/acme/client.py", line 1211, in _post_once
    data = self._wrap_in_jws(obj, self._get_nonce(url, new_nonce_url), url, acme_version)
  File "/tmp/cb/lib/python3.8/site-packages/acme/client.py", line 1030, in _wrap_in_jws
    return jws.JWS.sign(jobj, **kwargs).json_dumps(indent=2)
  File "/tmp/cb/lib/python3.8/site-packages/acme/jws.py", line 51, in sign
    return super(JWS, cls).sign(payload, key=key, alg=alg,
  File "/tmp/cb/lib/python3.8/site-packages/josepy-1.3.0.dev0-py3.8.egg/josepy/jws.py", line 267, in sign
    cls.signature_cls.sign(payload=payload, **kwargs),))
  File "/tmp/cb/lib/python3.8/site-packages/josepy-1.3.0.dev0-py3.8.egg/josepy/jws.py", line 229, in sign
    signature = alg.sign(key.key, cls._msg(protected, payload))
  File "/tmp/cb/lib/python3.8/site-packages/josepy-1.3.0.dev0-py3.8.egg/josepy/jwa.py", line 112, in sign
    return key.sign(msg, self.padding, self.hash)
TypeError: sign() takes 3 positional arguments but 4 were given
2020-12-31 11:52:55,570:ERROR:certbot._internal.log:An unexpected error occurred:
2020-12-31 11:52:55,570:ERROR:certbot._internal.log:TypeError: sign() takes 3 positional arguments but 4 were given
```

**Comment by commonism** [comment]
Using the mentioned josepy version, the following patches solves the problem:

```diff
diff --git a/acme/acme/client.py b/acme/acme/client.py
index d413ce13d..377f41f1c 100644
--- a/acme/acme/client.py
+++ b/acme/acme/client.py
@@ -1016,8 +1016,9 @@ class ClientNetwork(object):
             obj.le_acme_version = acme_version
         jobj = obj.json_dumps(indent=2).encode() if obj else b''
         logger.debug('JWS payload:\n%s', jobj)
+        alg = {'RSA':jose.RS256,'EC':jose.ES256}.get(self.key.typ)
         kwargs = {
-            "alg": self.alg,
+            "alg": alg,
             "nonce": nonce
         }
         if acme_version == 2:
```

**Comment by alexzorin** [comment]
Thanks! Until this is fixed, users can still  revoke ECDSA certs using `--cert-name`, which will sign the revocation using the account key.

**Comment by osirisinferi** [comment]
I can't seem to reproduce this on current certbot from git (version 1.11.0.dev0)?

```
server ~ # certbot revoke --staging --cert-path /etc/letsencrypt/live/le-test-04.example.com/cert.pem 
Saving debug log to /var/log/letsencrypt/letsencrypt.log

Please choose an account
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
1: server@2017-02-27T20:01:26Z (9372)
2: server@2017-03-13T13:26:15Z (26cc)
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Select the appropriate number [1-2] then [enter] (press 'c' to cancel): 1

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Would you like to delete the cert(s) you just revoked, along with all earlier
and later versions of the cert?
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
(Y)es (recommended)/(N)o: n
Congratulations! You have successfully revoked the certificate that was located at /etc/letsencrypt/live/le-test-04.example.com/cert.pem.
server ~ # 
```

(It was of course an ECDSA certificate:

```
  Certificate Name:   le-test-04.example.com
    Serial Number:    facc72830fc7d5b3724204f63a675f522a44
    Key Type:         ECDSA
    Domains:          le-test-04.example.com
    Expiry Date:      2021-04-02 17:42:00+00:00 (INVALID: TEST_CERT, REVOKED)
    Certificate Path: /etc/letsencrypt/live/le-test-04.example.com/fullchain.pem
    Private Key Path: /etc/letsencrypt/live/le-test-04.example.com/privkey.pem
```
)

Could this be due to your installation through pip? Which, as far as I know, isn't recommended. Although I'm not sure if it's also unsupported.

**Comment by bmw** [comment]
This was fixed in #8725.

**Comment by JavaScriptDude** [comment]
Note for Ubuntu users - apt installs a rather old version of `certbot` (1.21.0) which still has this bug. I found that `snap` installed the latest version and had no issues.
